# Assignment 2: Obesity Diagnosis based on eating Habits and physical condition

Student Name: Oliver Quarm
 
Student ID: 834844

#### <b>Due date</b>: Monday, 6 September 2021 5pm

<b>Submission method</b>: Canvas submission

<b>Submission materials</b>: completed copy of this iPython notebook

<b>Late submissions</b>: -10% per day (both week and weekend days counted up to 5 days)
<ul>
    <li>one day late, -1.5;</li>
    <li>two days late, -3.0;</li>
    <li>three days late, -4.5;</li>
    <li>four days late, -6.0;</li>
    <li>five days late, -7.5;</li>
</ul>

<b>Marks</b>: 15% of mark for class. 


<b>Materials</b>: See [Using Jupyter Notebook and Python page](https://canvas.lms.unimelb.edu.au/courses/105477/pages/python-and-jupyter-notebooks?module_item_id=2613813) on Canvas (under Modules> Coding Resources) for information on the basic setup required for this class, including an iPython notebook viewer and the python packages Numpy, Scipy, Scikit-Learn. The instructions for each question will state what libraries you can use; if your iPython notebook doesn't run on the marker's machine, you will lose marks. <b> You should use Python 3</b>.  


<b>Evaluation</b>: Your iPython notebook should run end-to-end without any errors in a reasonable amount of time, and you must follow all instructions provided below, including specific implementation requirements and instructions for what needs to be printed (please avoid printing output we don't ask for). You should edit the sections below where requested, but leave the rest of the code as is. You should leave the output from running your code in the iPython notebook you submit, to assist with marking. The amount each section is worth is given in parenthesis after the instructions. 

You should be careful to use Python built-in functions and operators when appropriate and pick descriptive variable names that adhere to <a href="https://www.python.org/dev/peps/pep-0008/">Python style requirements</a>. If you think it might be unclear what you are doing, you should comment your code to help the marker make sense of it. While the main focus is on correctness of your methods, you will lose marks if your code is not understandable.

<b>Updates</b>: Any major changes to the assignment will be announced via Canvas. Minor changes and clarifications will be announced on the discussion board; we recommend you check it regularly.

<b>Academic misconduct</b>: For most people, collaboration will form a natural part of the undertaking of this homework, and we encourage you to discuss it in general terms with other students. However, this ultimately is still an individual task, and so reuse of code or other instances of clear influence will be considered cheating. We will be checking submissions for originality and will invoke the University’s <a href="http://academichonesty.unimelb.edu.au/policy.html">Academic Misconduct policy</a> where inappropriate levels of collusion or plagiarism are deemed to have taken place.

## Overview

In this assignment, you will be working with *obesity* dataset containing data for the estimation of obesity levels in individuals from the countries of Mexico, Peru and Colombia, based on their eating habits and physical condition. The data contains 16 attributes describing 2111 patients. The dataset provided for this assignment is derived from <a href="https://archive.ics.uci.edu/ml/datasets/Estimation+of+obesity+levels+based+on+eating+habits+and+physical+condition+#">this</a> resource. Your task is to predict whether a patient is considered obese (mass body index >=30) or not (mass body index <30).

The attributes are

|ID|Feature Name| Feature Type | Feature Values|
| :-| :-| :-| :-|
|0|Gender| categorical | Female: 0, Male: 1|
|1|Age| continuous | |
|2|Height| continuous | |
|3|Weight| continuous | ? |
|4|family_history| categorical | yes: 1, no: 0|
|5|FACV| categorical | yes: 1, no: 0|
|6|FCVC| continuous | |
|7|NCP| continuous | |
|8|CAEC| categorical | Always: 0, Frequently: 1, Sometimes: 2, no:  3|
|9|SMOKE| categorical | yes: 1, no: 0|
|10|CH2O| continuous | |
|11|SCC| categorical | yes: 1, no: 0|
|12|FAF| continuous | |
|13|TUE| continuous | |
|14|CALC| categorical | Always: 0, Frequently: 1, Sometimes: 2, no:  3, ?|
|15|MTRANS| categorical | Automobile: 0, Bike: 1, Motorbike: 2, Public_Transportation: 3, Walking: 4, ?|

You can find out more about the individual attributes / values, and the origin of the data set in this <a href="https://www.sciencedirect.com/science/article/pii/S2352340919306985"> article</a>  


You will build a number of classifiers to predict whether a patient is obese or not based on the attributes above.


#### The following instructions hold for every question in the assignment
- leave the order of instances intact, i.e., do not shuffle the data
- do not change the names or types of variables provided by us in the code cells below
- '?' denotes a missing value, and you should treat it as instructed in Question 1-B.
- the mapping of categorical feature values to integers indicates how you should process your data in 1-C.

## Question 1: Loading and pre-processing the data [2 marks]

You were provided with two data files:

**obesity.train** contains about 1,688 training instances, one instance per line in comma-separated value (csv) format. Each line contains 17 fields. The first 16 fields correspond to the features listed above, the final field denotes the class label

**obesity.test** is formatted exactly like obesity.train, and contains about 423 further instances for evaluation.

### 1-A Read the data [0.25 marks]

First, you will read in the data and create train features, train labels, test features and test labels. Do not apply any data transformations in this step and keep the '?' as is. 

In [1]:
# load data

data = open("obesity.train",'r').readlines()
test_data = open("obesity.test",'r').readlines()
feature_names = ["gender","age","height","weight","family_history","FACV","FCVC","NCP","CAEC","smoke","CH2O","SCC","FAF","TUE","CALC","MTRANS"]

x_train = []
y_train = []
x_test = []
y_test = []

############################
## your code begins here
############################

for i in data:
    row = i.split(",")
    row[-1] = row[-1].strip()
    rows_features = []
    for j in range(0,len(row)-1):
        rows_features.append(row[j])
    x_train.append(rows_features)
    y_train.append(row[-1])

for i in test_data:
    row = i.split(",")
    row[-1] = row[-1].strip()
    rows_features = []
    for j in range(0,len(row)-1):
        rows_features.append(row[j])
    x_test.append(rows_features)
    y_test.append(row[-1])

############################
## your code ends here
############################

len(x_train), len(y_train), len(x_test), len(y_test), len(x_train[0]), len(x_test[0])

(1688, 1688, 423, 423, 16, 16)

<b>For your testing:</b>

In [2]:
assert len(x_train)==len(y_train)==1688
assert len(x_train[0])==len(x_test[0])==16
assert len(x_test)==len(y_test)==423
assert x_train[5][2] == '1.7544389999999999'
assert x_test[8][6] == '2.0'

### 1-B: Dealing with Missing Values [0.75 marks]

You will write <b>two functions</b> to replace missing values using the strategies below based on the different attribute types (categorical, continuous):

   * <b>Continuous features</b>: For each feature find the <b>average feature value</b> in the train dataset 
   * <b>Categorical features</b>: For each feature find the <b>most frequent value</b> in the train dataset 

and a <b>third function</b> that replaces the missing values in the train and test datasets with the most frequent (categorical features) and average values (continuous features) that you found in the train dataset.
 

**Your tasks**

You will implement three functions
1. `num_mean_feat`
<ul>
  <li>input:  instances with only continuous features</li>
  <li>output: a list containing the average value for each continuous feature (aka substitute values)</li>
</ul> 
    
2. `cat_most_feat`
<ul>
    <li>input: instances with only categorical features</li>
    <li>output: a list containing the most frequent feature value for each categorical feature (aka substitute values)</li>
</ul>

3. `transform` 
<ul>
    <li>input:
        <ul>
            <li>a list of instances</li>   
            <li>replacement list containing the substitute values for each feature</li>
        </ul>
    </li>
    <li>output: the transformed instances</li>
</ul>

<b> You should implement these functions from scratch yourself. You may use native Python libraries such as math, collections, or numpy to help you, but you may not use existing implementations.</b>

In [3]:
import copy
from collections import Counter
import numpy as np
from numpy.core.fromnumeric import shape, size
from sklearn import preprocessing

def num_mean_fit(X):
    replacement = []
    np_array = np.array(X)
    no_of_instances = len(np_array)
    no_of_features = len(np_array[0])
    for i in range(0, no_of_features):            # Number of features.
        features_added = 0
        total_nonmissing_values = 0
        for j in range(0, no_of_instances):       # Number of instances.
            try:
                float(np_array[j][i])
            except ValueError:
                continue
            features_added += float(np_array[j][i])
            total_nonmissing_values += 1
        average_feature = features_added/total_nonmissing_values
        replacement.append(average_feature)
    return replacement

def cat_most_fit(X):
    replacement = []
    np_array = np.array(X)
    no_of_instances = len(np_array)
    no_of_features = len(np_array[0])
    for i in range(0, no_of_features):            # Number of features.
        category_values = []
        for j in range(0, no_of_instances):       # Number of instances.
            category_values.append(np_array[j][i])
        occurence_count = Counter(category_values)
        winning_category = occurence_count.most_common(1)[0][0]
        replacement.append(winning_category)
    return replacement

def transform(X, replacement):
    trans_X  = copy.deepcopy(X)
    np_array = np.array(X)
    no_of_instances_to_correct = len(X)
    no_of_features_to_check = len(np_array[0])
    for i in range(0, no_of_features_to_check):              # Number of features.
        for j in range(0, no_of_instances_to_correct):       # Number of instances.
            if np_array[j][i] == '?':
                replacement_value = replacement[i]
                trans_X[j][i] = replacement_value
    return trans_X

<b>For your testing:</b>

In [4]:
assert cat_most_fit([['no','always'],['yes','always'],['yes','?']]) == ['yes','always']
assert num_mean_fit([[12,122,'?'],[21,150,78],[18,'?',79]]) == [17,136,78.5]
assert transform([[10,144,'?'],[92,'?',70]], [17,136,78.5]) == [[10, 144, 78.5], [92, 136, 70]]

### 1-C: Attribute Types [1 marks]

You will create four feature representations, based on the different attribute types (categorical, continuous) in the original *Obesity* data.

**Your tasks**

Denote with $I$ the number of training instances; $C$ the number of categorical features in the data set; $N$ the number of continuous features; $v_{f}$ the number of values for categorical feature $f$; and $v_{bin}$ the number of bins for continuous features. 

1. Create a train data set with only numeric features `x_train_num` (size $(I\times N)$); and equivalently a test data set `x_test_num`; the missing values in train and test must be replaced with the average feature value found in the train dataset

2. Create a train data set with only categorical features `x_train_cat` (size $(I\times C)$); and equivalently a test data set `x_test_cat`; the missing values in train and test must be replaced with the most frequent feature values in the train dataset; You need to convert the feature values in `x_train_cat` and `x_test_cat` to integers (as shown in the Overview table) to be able to further use them in the rest of your program

3. Create a train data set with both continuous `x_train_num` and a 1-hot representation of categorical features `x_train_1hot` and save into a variable called `x_train_num_full` (size $I \times (\sum_f v_{f}+N)$) where the first $\sum_f v_f$ columns represent the 1-hot encoded categorical features and the remaining columns the continuous features; and equivalently a test data set `x_test_num_full`

4. Create a train data set using both categorical features `x_train_cat` and discretized numeric features `x_train_disc`(use equal-width binning with 4 bins, $v_{bin}=4$) and save into a variable called `x_train_cat_full` (size $I\times (C+N)$); and equivalently a test data set `x_test_cat_full`; You need to convert the feature values in `x_train_cat_full` and `x_test_cat_full` to integers (as shown in the Overview table) to be able to further use them in the rest of your program.


You will create two lists of feature names for the full categorical and full continuous datasets.

1. Create a list of feature names (string) for the `x_{train,test}_num_full` dataset (size $C+N$).

2. Create a list of feature names (string) for the `x_{train,test}_cat_full` dataset (size $\sum_f v_{f}+N$).


**Note:** You may use classes and functions from ```scikit-learn```. You may use native Python libraries such as math, collections, or numpy to help you.

In [5]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

x_train_num = []
x_test_num = []

x_train_cat = []
x_test_cat = []

x_train_num_full = []
x_test_num_full = []

x_train_cat_full = []
x_test_cat_full = []

full_num_feature_names = []
full_cat_feature_names = []

numpy_array_train = np.array(x_train)
numpy_array_test = np.array(x_test)

num_train_precorrection = numpy_array_train[:, [1,2,3,6,7,10,12,13]] # Instances containing only numbers.
cat_train_precorrection = numpy_array_train[:, [0,4,5,8,9,11,14,15]]
num_test_precorrection = numpy_array_test[:, [1,2,3,6,7,10,12,13]]   # Instances containing only numbers.
cat_test_precorrection = numpy_array_test[:, [0,4,5,8,9,11,14,15]]

replacement_values_num = num_mean_fit(num_train_precorrection) # 1D array containing num replacement values.
replacement_values_cat = cat_most_fit(cat_train_precorrection) # 1D array containing cat replacement values.

x_train_num = transform(num_train_precorrection, replacement_values_num)
x_train_cat = transform(cat_train_precorrection, replacement_values_cat)
x_test_num = transform(num_test_precorrection, replacement_values_num)
x_test_cat = transform(cat_test_precorrection, replacement_values_cat)

# The missing values have now all been replaced and corrected for. Below we are going to 1-Hot encode the categorical
# values (x_train_1hot), and then concatenate this with the continuous data to form x_{train/test}_num_full.

columnTransformer = ColumnTransformer([('encoder', OneHotEncoder(), [0,1,2,3,4,5,6,7])], remainder='passthrough')
columnTransformer.fit(x_train_cat)
x_train_1hot = np.array(columnTransformer.transform(x_train_cat), dtype = str)
x_train_num_full = np.concatenate((x_train_1hot, x_train_num),axis=1)

x_test_1hot = np.array(columnTransformer.transform(x_test_cat), dtype = str)
x_test_num_full = np.concatenate((x_test_1hot, x_test_num),axis=1)

cat_train_integer  = copy.deepcopy(x_train_cat)
cat_test_integer  = copy.deepcopy(x_test_cat)

# Now we're going to LabelEncode the categorical data to integers, which we will then concatenate with the discretized 
# continuous data to thus form x_{train/test}_cat_full.

for i in range(0, 8):
    le = preprocessing.LabelEncoder()
    cat_train_integer[:, i] = le.fit_transform(cat_train_integer[:, i])
    cat_test_integer[:, i] = le.transform(cat_test_integer[:, i])

num_train_binned = copy.deepcopy(x_train_num)
num_test_binned = copy.deepcopy(x_test_num)
num_train_binned = num_train_binned.astype('float64')
num_test_binned = num_test_binned.astype('float64')

age_max, height_max, weight_max, FCVC_max, NCP_max, CH2O_max, FAF_max, TUE_max = num_train_binned.max(axis=0)
age_min, height_min, weight_min, FCVC_min, NCP_min, CH2O_min, FAF_min, TUE_min = num_train_binned.min(axis=0)
max_values = [age_max, height_max, weight_max, FCVC_max, NCP_max, CH2O_max, FAF_max, TUE_max]
min_values = [age_min, height_min, weight_min, FCVC_min, NCP_min, CH2O_min, FAF_min, TUE_min]

# The below function discretizes the continuous data into bins.
def discrete_binner(array):
    for i in range(0, 8):
        bin_size = (max_values[i] - min_values[i])/4
        bin_zero_lowerbound = min_values[i] 
        bin_one_lowerbound = min_values[i] + bin_size
        bin_two_lowerbound = min_values[i] + (2*bin_size)
        bin_three_lowerbound = min_values[i] + (3*bin_size)
        for j in range(0, len(array)):
            if bin_zero_lowerbound <= array[j][i] < bin_one_lowerbound:
                array[j][i] = 0
            elif bin_one_lowerbound <= array[j][i] < bin_two_lowerbound:
                array[j][i] = 1
            elif bin_two_lowerbound <= array[j][i] < bin_three_lowerbound:
                array[j][i] = 2
            elif bin_three_lowerbound <= array[j][i] <= max_values[i]:
                array[j][i] = 3
            else:
                raise Exception('Strange number.')
    array = array.astype('int32')
    return array

num_train_binned = discrete_binner(num_train_binned)
num_test_binned = discrete_binner(num_test_binned)

x_train_cat_full = np.concatenate((cat_train_integer, num_train_binned),axis=1)
x_test_cat_full = np.concatenate((cat_test_integer, num_test_binned),axis=1)

full_cat_feature_names = ['Gender_int', 'Fam_History_int', 'FACV_int', 'CAEC_int', 'SMOKE_int', 'SCC_int', 'CALC_int', 'MTRANS_int', 'Age_bin', 'Height_bin', 'Weight_bin', 'FCVC_bin', 'NCP_bin', 'CH2O_bin', 'FAF_bin', 'TUE_bin']
full_num_feature_names = ['Gender_Female', 'Gender_Male', 'Fam_History_No', 'Fam_History_Yes', 'FACV_No', 'FACV_Yes', 'CAEC_Always', 'CAEC_Frequently', 'CAEC_Sometimes', 'CAEC_no', 'SMOKE_No', 'SMOKE_Yes', 'SCC_No', 'SCC_Yes', 'CALC_Always', 'CALC_Frequently', 'CALC_Sometimes', 'CALC_no',  'MTRANS_Automobile', 'MTRANS_Bike', 'MTRANS_Motorbike', 'MTRANS_Public_Transportation', 'MTRANS_Walking', 'Age', 'Height', 'Weight', 'FCVC', 'NCP', 'CH2O', 'FAF', 'TUE']

# Convert sets into their appropriate datatypes for use later in my program.
x_train_num_full = x_train_num_full.astype('float64')
x_test_num_full = x_test_num_full.astype('float64')
x_train_num = x_train_num.astype('float64')
x_test_num = x_test_num.astype('float64')
x_train_cat_full = x_train_cat_full.astype('int32')
x_test_cat_full = x_test_cat_full.astype('int32')

<b>For your testing:</b>

In [6]:
assert len(x_train_cat[0])==8
assert len(x_train_num[0])==8
assert len(x_train_cat_full[0])==16
assert len(x_train_num_full[0])==31

## Question 2: A Weighted Random baseline [1.25 marks]

Implement a <b>weighted random</b> baseline that randomly assigns a class to each test instance, weighting the class assignment according to <b>class prior probabilities</b>, as introduced in the Evaluation lecture and discuss your results.

### 2-A: Implement a weighted random baseline [0.5 marks]

**Your tasks**

You will implement the baseline function
1. `weight_random`
<ul>
  <li>input:
      <ul>
          <li> test labels</li>
          <li> the prior probability of class "obese" </li>          
      </ul>
  </li>
  <li>output: error rate on the test (fraction of incorrect predictions)</li>          
</ul> 
    

<b> You should implement this function from scratch yourself. We have already imported the two functions that we assume you will need. Do not import any other function or library.</b>

In [7]:
from collections import Counter
import collections
from random import random
import random

test_obese = 0  ## calculate the fraction of obese patients in the test set 
prior_obese = 0 ## calculate the prior probability of class obese in the train set

def obese_fraction(array, obese_counter):
    for i in range(0, len(array)):
        if array[i] == 'obese':
            obese_counter += 1
    return obese_counter

test_ObeseLabelCount = 0
test_ObeseLabelCount = obese_fraction(y_test, test_ObeseLabelCount)
train_ObeseLabelCount = 0
train_ObeseLabelCount = obese_fraction(y_train, train_ObeseLabelCount)

test_obese = test_ObeseLabelCount/len(y_test)    ## calculate the fraction of obese patients in the test set
print(f'Test obese fraction: {test_obese}')
prior_obese = train_ObeseLabelCount/len(y_train) ## calculate the prior probability of class obese in the train set
print(f'Prior/train obese fraction: {prior_obese}\n')

def weight_random(y_test, prior_obese):
    error = 0
    correct_counter = 0
    label_options = ['obese', 'notObese']    
    c = collections.Counter()
    for i in range(0, len(y_test)):
        chosen_label = random.choices(label_options, weights=(prior_obese, 1-prior_obese))
        if y_test[i] == chosen_label[0]:
            correct_counter += 1
    error = 1-(correct_counter/len(y_test))
    return error 

err = weight_random(y_test, prior_obese)
print("There are",test_obese*len(y_test),"obese and",(1-test_obese)*len(y_test),"non-obese test instances")
print("The prior probability of obese is", round(prior_obese,2))
print("The weighted random baseline's error is:", round(err,2))

##########################################
## insert code for for Question 2-B here:
##########################################

error_values = []
for i in range(0, 1000):
    value = weight_random(y_test, prior_obese)
    error_values.append(value)
print(f'\n---Part B---\n\nThe average error over 1000 iterations is: {(sum(error_values))/len(error_values)}')

print(f'The error rate should converge to: {1 - ((prior_obese*test_obese)+(1-prior_obese)*(1-test_obese))}')

Test obese fraction: 0.47044917257683216
Prior/train obese fraction: 0.45793838862559244

There are 199.0 obese and 224.0 non-obese test instances
The prior probability of obese is 0.46
The weighted random baseline's error is: 0.52

---Part B---

The average error over 1000 iterations is: 0.49715839243498794
The error rate should converge to: 0.497514089162269


### 2-B: Weighted Random baseline Discussion [0.75 marks]

Repeat the random baseline 1000 times, what would be the <b>average error rate</b> of your weighted random baseline on the test dataset? You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 2B.

What would the error rate converge to (write a formula based on the prior probability of obese, `prior_obese`, and the fraction of obese samples in the test data, `test_obese`)?

Please limit your answer to 2-3 sentences

Q2-B) As shown via the print out, the average error over 1,000 iterations is approximately 0.49xxx, with the x's varying (performing more iterations reduces this variance). Given `prior_obese = a` and `test_obese = b`, the error rate should converge to: `1 - ((a*b) + (1-a)*(1-b)) = 0.49751`. 

## Question 3: One-R baseline [2.5 marks]

In this question you will implement the one-R baseline using a helper function and discuss your results. Please refer to the Evaluation lecture for One-R baseline. 

### Question 3-A: Implement a helper function [0.75 mark]

**Your tasks**

You will implement a function that calculates the error rate for each attribute if we select that attribute as the one rule for classification.
1. `feature_error`
<ul>
  <li>input:
      <ul>
          <li> features</li>
          <li> labels</li>
      </ul>
  </li>
  <li>output:
      <ul>
          <li> a dictionary of the form {feature: error rate on the train} where each 
        feature is denoted by its ID (position in the train dataset)</li>
          <li> a dictionary of the form {feature value: majority class label} where the majority class label is either "obese" or "notObese" and feature values are the best feature's possible values (feature with smallest error rate)</li>
      </ul>
    </li>   
</ul> 

<b> You should implement this function from scratch yourself. You may use native Python libraries like math, numpy, collections to help you, but you may not use existing implementations.</b>


In [8]:
# Takes in a dictionary for one features value. Returns, for that value, which label was seen the most, the total count of that label, 
# and how many times that value is present in that feature column.
def dict_values_total(dictionary):
    total = 0
    for value in dictionary.values():
        total += value
    all_values = dictionary.values()
    max_value = max(all_values)
    max_key = max(dictionary, key=dictionary.get)
    return max_key, max_value, total

# Main function.
def feature_error(features, labels):
    feat_error = {}
    feature_dictionaries = {}  
    feat_rule = {}
    label_options = list(set(labels))
    label_options_dict = dict.fromkeys(label_options, 0)
    for i in range(0, len(features[0])):
        # Need to make a dictionary for each value in this feature. 
        value_dictionaries = {}
        for j in range(0, len(features)):
            value = features[j][i]
            if value not in value_dictionaries:
                a = copy.deepcopy(label_options_dict)
                value_dictionaries[value] = a
            label = labels[j]
            value_dictionaries[value][label] += 1
        feature_dictionaries[i] = value_dictionaries
        total_count = 0
        total_correct = 0
        for key in value_dictionaries:
            max_key, max_value, total = dict_values_total(value_dictionaries.get(key))
            total_count += total
            total_correct += max_value

        total_wrong = total_count - total_correct
        feature_total_errors = total_wrong/total_count
        feat_error[i] = feature_total_errors

    best_feature = min(feat_error, key=feat_error.get)
    best_value_dictionaries = feature_dictionaries.get(best_feature)
    for key in best_value_dictionaries:
        dictionary = best_value_dictionaries.get(key)
        most_frequent_label = max(dictionary, key=dictionary.get)
        feat_rule[key] = most_frequent_label
    return feat_error, feat_rule

<b>For your testing</b>

In [9]:
sample_features = [[2,2], [2,1], [4,2], [4,1]]
sample_labels = ["yes","yes","no","no"]
sample_error, sample_rule= feature_error(sample_features, sample_labels)

# where index 0 refers to "feature 1" and index 1 to "feature 2"
assert sample_error[0]==0.0
assert sample_error[1]==0.5

# where 2 and 4 refer to the values of the best feature 
# and "yes","no" to the majority class label
assert sample_rule == {2:"yes",4:"no"}

### Question 3-B: Implement One-R baseline [0.75 mark]

- Apply your `feature_error` function to the full training data *x_train_cat_full*, and determine the (i) feature with *highest* error rate and (ii) feature with *lowest* error rate. Store the name (string) of the corresponding highest/lowest error rate features in `highest_err_feature_name` and `lowest_err_feature_name`, respectively.

- The feature with *lowest* error will constitute your 1-R predictor, which you should use to predict the class labels for the *x_test_cat_full* (`one_r_predictions`)


In [10]:
one_r_predictions = []

highest_err_feature_name = "" # feature name with highest error
lowest_err_feature_name = ""  # feature name with lowest error

############################
## your code begins here
############################

x_train_cat_full_errors, x_train_cat_full_rule = feature_error(x_train_cat_full, y_train)
best_feature = min(x_train_cat_full_errors, key=x_train_cat_full_errors.get)
worst_feature = max(x_train_cat_full_errors, key=x_train_cat_full_errors.get)
highest_err_feature_name = full_cat_feature_names[worst_feature]
lowest_err_feature_name = full_cat_feature_names[best_feature]
for i in range(0,len(x_test_cat_full)):
    if x_test_cat_full[i][10] == '0' or x_test_cat_full[i][10] == '1':
        one_r_predictions.append('notObese')
    else:
        one_r_predictions.append('obese')

############################
## your code ends here
############################

print(one_r_predictions[:10])
print(f"The feature with highest error is: {highest_err_feature_name}")
print(f"The feature with lowest error is: {lowest_err_feature_name}")

##########################################
## insert code for for Question 3-D here:
##########################################

print(f'\n---Part D---\n\nAll 16 1-R scores (one 1-R score per feature) are as follows:\n\n{x_train_cat_full_errors}')

['obese', 'obese', 'obese', 'obese', 'obese', 'obese', 'obese', 'obese', 'obese', 'obese']
The feature with highest error is: Gender_int
The feature with lowest error is: Weight_bin

---Part D---

All 16 1-R scores (one 1-R score per feature) are as follows:

{0: 0.45793838862559244, 1: 0.3684834123222749, 2: 0.4490521327014218, 3: 0.3921800947867299, 4: 0.45734597156398105, 5: 0.45793838862559244, 6: 0.4443127962085308, 7: 0.45793838862559244, 8: 0.35545023696682465, 9: 0.43838862559241704, 10: 0.2334123222748815, 11: 0.41824644549763035, 12: 0.39514218009478674, 13: 0.4206161137440758, 14: 0.42890995260663506, 15: 0.37618483412322273}


### Question 3-C: Compute Mutual Information (MI) [0.5 mark]

You will calculate the mutual information between your features  *x_train_cat_full* and the class label and determine the (i) feature with *highest* MI and (ii) feature with *lowest* MI. Store the name (string) of the corresponding highest/lowest MI features in `highest_mi_feature_name` and `lowest_mi_feature_name`, respectively.

**Note:** You may use classes and functions from ```scikit-learn``` but make sure that you specify your dataset is categorical (discrete) otherwise it will be assumed that your data is numeric. 

In [11]:
highest_mi_feature_name = "" # feature with highest MI
lowest_mi_feature_name = "" # feature with lowest MI

############################
## your code begins here
############################

from sklearn.feature_selection import mutual_info_classif as MIC

mi_score = MIC(x_train_cat_full, y_train, discrete_features=True)

index_of_largest_mi = np.argmax(mi_score)
largest_mi = mi_score[index_of_largest_mi]
index_of_lowest_mi = np.argmin(mi_score)
lowest_mi = mi_score[index_of_lowest_mi]
highest_mi_feature_name = full_cat_feature_names[index_of_largest_mi]
lowest_mi_feature_name = full_cat_feature_names[index_of_lowest_mi]

############################
## your code ends here
############################

print(f"The feature with highest MI is: {highest_mi_feature_name}")
print(f"The feature with lowest MI is: {lowest_mi_feature_name}")

##########################################
## insert code for for Question 3-D here:
##########################################

print(f'\n---Part D---\n\nAll 16 MIC scores are as follows:\n{np.round(mi_score, 3)}')

The feature with highest MI is: Weight_bin
The feature with lowest MI is: Gender_int

---Part D---

All 16 MIC scores are as follows:
[0.    0.109 0.042 0.085 0.    0.022 0.014 0.013 0.04  0.016 0.357 0.013
 0.057 0.013 0.014 0.044]


### Question 3-D: Discussion of One-R and Mutual Information [0.5 mark]
In addition to features leading to lowest/highest error rates and MI, inspect the MI and error rates for all 16 features. You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 3D.  

Are you finding different or somewhat similar patterns in the results? Explain the difference and similarity between MI and error rate formula in One-R?

Please limit your answer to 2-3 sentences

Q3-D) In both Mutual Information and One-R, the `Weight` feature is demonstrated as having the strongest connection to the class labels. Also in both, the `Gender` feature is demonstrated as having the weakest connection to the class labels. However, on closer inspection, in just the One-R data, `Gender`, `SCC` and `MTRANS` all have the same highest error value -- it's not the case that solely `Gender` has the highest error!

## Question 4: Naive Bayes [2 marks]

We will construct two Naive Bayes classifiers

1. One for instances with categorical attributes.
2. One for instances with numerical attributes.


For each classifier, you will 
1. Train it on the training set
2. Use the models to predict labels for the test set

### Question 4-A: Implementing the Naive Bayes classifiers [0.5 marks]

Implement two functions which train a NB classifier given the specified input feature types and predict labels for a given test set.

You may add additional quantities to your `return` statements.

<b> You may (and are, indeed, encouraged) to, use the existing NB implementations from `sklearn`. You should use the default parameterizations of these algorithms.</b> 
    
**Note:** If you choose to implement your classifiers from scratch, please use Laplace smoothing (alpha=1) for the categorical feature NB.




In [12]:
##################################
## your code begins here
##################################

from sklearn.naive_bayes import CategoricalNB, GaussianNB

def nb_cat_features(train_features, train_labels, test_features):
    predictions = []      
    cnb = CategoricalNB()
    cnb.fit(train_features, train_labels)
    predictions = cnb.predict(test_features)
    return predictions

def nb_num_features(train_features, train_labels, test_features):
    predictions = []
    gnb = GaussianNB()
    gnb.fit(train_features, train_labels)
    predictions = gnb.predict(test_features)
    return predictions

############################
## your code ends here
############################

### Question 4-B: Apply your classifiers to the data sets you created in Questions 1 [0.5 marks]

...namely, the

1. categorical features `x_{train,test}_cat`
2. numerical features `x_{train,test}_num`
3. combined discretized numerical and categorical features `x_{train,test}_cat_full`
4. combined 1-hot categorical features and numerical `x_{train,test}_num_full` 


In [13]:
from collections import Counter

############################
## your code begins here
############################

cat_nb_predictions = nb_cat_features(cat_train_integer, y_train, cat_test_integer) # predictions for x_test_cat
num_nb_predictions = nb_num_features(x_train_num, y_train, x_test_num) # predictions for x_test_num
full_cat_nb_predictions = nb_cat_features(x_train_cat_full, y_train, x_test_cat_full) # predictions for x_test_cat_full
full_num_nb_predictions = nb_num_features(x_train_num_full, y_train, x_test_num_full) # predictions for x_test_num_full

############################ 
## your code ends here
############################

print(f"Numerical NB predicted class distribution \t {Counter(num_nb_predictions)}")
print(f"Categorical NB predicted class distribution\t {Counter(cat_nb_predictions)}")
print(f"Full Categorical NB predicted class distribution\t {Counter(full_cat_nb_predictions)}")
print(f"Full Numerical NB predicted class distribution\t {Counter(full_num_nb_predictions)}")

Numerical NB predicted class distribution 	 Counter({'notObese': 237, 'obese': 186})
Categorical NB predicted class distribution	 Counter({'obese': 261, 'notObese': 162})
Full Categorical NB predicted class distribution	 Counter({'notObese': 215, 'obese': 208})
Full Numerical NB predicted class distribution	 Counter({'obese': 261, 'notObese': 162})


### Question 4-C: Combining NB predictions [1 mark]

You are asked to construct another Naive Bayes classifier for instances with the full set of numerical *and* categorical attributes (no discretization and no one-hot encoding), ensuring that your classifier computes posterior class probabilities $p(y|x)$ as $p(x|y)p(y)$. Explain such Naive Bayes classifier implementation on the full attribute set (numerical and categorical features) 

Please limit your answer to 3-4 sentences

Q4-C) Such an implementation containing both numerical and categorical features can be done. First, for all columns, we construct the frequency & likelihood tables for all features. The likelihood for the categorical features is calculated with the help of the tables as: 𝑝(𝑦|𝑥) = 𝑝(𝑥|𝑦)𝑝(𝑦). For the continuous feature columns, to calculate the likelihood, we first find the mean and standard deviation for each label (so if there's 2 possible labels, there will be 2 sets of mean and standard deviation). We then use the mean and standard deviation to calculate the normal distribution value for each label, which is its likelihood. The likelihoods across all columns are then multiplied together (per label possibility), and the label with the largest value is the selected output label for that instance.

## Question 5: Logistic Regression [1.25 marks]

1. First find the mean and standard deviation of values for each feature in `x_train_num_full` and then using the calculated mean and standard deviation values, standardize `x_{train,test}_num_full` such that each feature has mean=0 and std= 1. Store the standardized data to `x_{train,test}_norm_full`.

2. Implement a Logistic Regression classifier, and apply it to the standardized full numeric training data set (`x_{train,test}_norm_full`).

<b> You should use existing implementation in sklearn with default parameters for Logistic Regression. You may use classes and functions from ```scikit-learn``` for standardization.</b>

### Question 5-A: The Logistic Regression classifier [0.5 marks]


In [14]:
from sklearn.linear_model import LogisticRegression

lr_predictions = []
x_train_norm_full = []
x_test_norm_full = []

############################
## your code begins here
############################

mean = np.mean(x_train_num_full, axis=0)
std = np.std(x_train_num_full, axis=0)
np.set_printoptions(suppress=True)

def normalizer(array):
    normalized_array = copy.deepcopy(array)
    for i in range(0, len(array[0])):
        for j in range(0, len(array)):\
            normalized_array[j][i] = (array[j][i] - mean[i])/std[i]  
    return normalized_array

x_train_norm_full = normalizer(x_train_num_full)
x_test_norm_full = normalizer(x_test_num_full)

logisticRegr = LogisticRegression()
logisticRegr.fit(x_train_norm_full, y_train)
lr_predictions = logisticRegr.predict(x_test_norm_full)
score = logisticRegr.score(x_test_norm_full, y_test)
print(f'The accuracy of this LR is: {100*round(score, 3)}%.\n')

############################
## your code ends here
############################

print(f"LR predicted class distribution\t {Counter(lr_predictions)}")

##########################################
## insert code for for Question 5-B here:
##########################################

print(f'\nThe coefficients for this LR classifier are:\n{logisticRegr.coef_}')

The accuracy of this LR is: 94.1%.

LR predicted class distribution	 Counter({'notObese': 227, 'obese': 196})

The coefficients for this LR classifier are:
[[ 0.12376967 -0.12376967 -0.3863544   0.3863544  -0.26233058  0.26233058
   0.0326411  -0.4583399   0.43143688 -0.13270847 -0.01516063  0.01516063
   0.25722798 -0.25722798 -0.00099834  0.04968252 -0.01572915 -0.00232277
  -0.02429057 -0.05939837  0.08813016  0.09764412 -0.21274504  0.03613459
  -1.67570728  6.40739956  0.12305897  0.11313373  0.07659881  0.00014641
  -0.14792405]]


### Question 5-B: Inspect the learned coefficients [0.75 marks]

**Part 1**
Inspect the coefficients (weights) learned by the classifier, and compare them to the features with the highest MI values in Question 3. You should update your code above in order to achieve this. Please indicate through comments which parts of the code are relevant to question 5B.

**Part 2**
What does a high coefficient for a feature imply? Report the two feature names for which your model learns the most important coefficients? Compare your answer to the highest MI features you detected in Q3.

Q5-B) A higher magnitude coefficient (positive or negative) implies a strong connection between a feature and the label. A stronger connection implies that this coefficient is important. Looking at the coefficients above, the two most important features are index values `25` (`-1.676`) and 26 (`6.41`), which correspond to `Height` and `Weight` respectively. `Weight` was noted in Q3 as having the strongest MI connection (which correlates with the notable weighted coefficient score noted just above), however `Height` does not exhibit a notable MI score. 

## Question 6: Evaluation [1 mark]

We will evaluate our baselines and classifiers on the instances in the test set. 

Compute 
- error rate
- macro-averaged F1 score 

for the One-R baselines, the four NB classifiers (on 4 datasets) and the LR classifier.

**You may use existing implementations and/or Python libraries like numpy, scipy or sklearn.**

In [15]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

one_r_err = 0
one_r_f1 = 0


num_nb_err = 0
num_nb_f1 = 0


cat_nb_err = 0
cat_nb_f1 = 0


full_cat_nb_err = 0
full_cat_nb_f1 = 0


full_num_nb_err = 0
full_num_nb_f1 = 0

lr_err = 0
lr_f1 = 0

############################
## your code begins here
############################

lr_err = 1 - accuracy_score(lr_predictions, y_test)
lr_f1 = f1_score(lr_predictions, y_test, average='macro')

one_r_err = 1 - accuracy_score(one_r_predictions, y_test)
one_r_f1 = f1_score(one_r_predictions, y_test, average='macro')

cat_nb_err = 1 - accuracy_score(cat_nb_predictions, y_test)
cat_nb_f1 = f1_score(cat_nb_predictions, y_test, average='macro')

num_nb_err = 1 - accuracy_score(num_nb_predictions, y_test)
num_nb_f1 = f1_score(num_nb_predictions, y_test, average='macro')

full_cat_nb_err = 1 - accuracy_score(full_cat_nb_predictions, y_test)
full_cat_nb_f1 = f1_score(full_cat_nb_predictions, y_test, average='macro')

full_num_nb_err = 1 - accuracy_score(full_num_nb_predictions, y_test)
full_num_nb_f1 = f1_score(full_num_nb_predictions, y_test, average='macro')

############################
## your code ends here
############################

print(f"One  R\t\tError: {round(one_r_err, 2)}\tMacro F1: {round(one_r_f1, 2)}")
print(f"NB Num \t\tError: {round(num_nb_err, 2)}\tMacro F1: {round(num_nb_f1, 2)}")
print(f"NB Cat \t\tError: {round(cat_nb_err, 2)}\tMacro F1: {round(cat_nb_f1, 2)}")
print(f"NB Num Full \tError: {round(full_num_nb_err, 2)}\tMacro F1: {round(full_num_nb_f1, 2)}")
print(f"NB Cat Full \tError: {round(full_cat_nb_err, 2)}\tMacro F1: {round(full_cat_nb_f1, 2)}")
print(f"LR \t\tError: {round(lr_err, 2)}\tMacro F1: {round(lr_f1, 2)}")

One  R		Error: 0.53	Macro F1: 0.32
NB Num 		Error: 0.13	Macro F1: 0.87
NB Cat 		Error: 0.21	Macro F1: 0.79
NB Num Full 	Error: 0.21	Macro F1: 0.79
NB Cat Full 	Error: 0.19	Macro F1: 0.81
LR 		Error: 0.06	Macro F1: 0.94


## Question 7: Discussion [5 marks] 

Critically analyze the performance of the models by answering the following questions. 

**(a)** The four Naive Bayes (NB) classifiers lead to different performance. Which of the four NB classifiers performs best, and why do you think it is the case? **[1 mark]**

**(b)** You used unstandardized data in your numeric Naive Bayes, what would be the effect of standardization on the results of numeric Naive Bayes? Why? **[0.75 mark]**

**(c)** You chose to use equal-width binning for your continuous features, discuss the drawbacks of considering this approach with a fixed bin number. Would you select a different binning strategy, and if so which? **[1 mark]**

**(d)** Assume that your best classifier will be deployed in a hospital to warn new patients if they are prone to obesity. You want to avoid falsely diagnosing a patient as obese. Can you reassure the hospital with the evaluation results in Q7 that your classifier is adequate? If not, describe an alternative evaluation metric and explain why it is more appropriate.**[0.75 mark]**

**(e)** Do you observe a clear difference in performance between NB and Logistic Regression (LR) in Q7? Referring to the number of parameters and assumptions underlying each model, provide one reason why NB might outperform LR, and one reason why LR might outperform NB. In your answer, refer back to the *Obesity* data set used in this assignment. **[1 mark]**

**(f)** Assume that you have access to a (hypothetical) infinitely large database (i.e., an enormous version of the Obesity data set used in this assignment). Is it true that Logistic Regression will achieve perfect test set performance? Why (not)? **[0.5 mark]**


<b>We expect a maximum of 2-3 sentences per question.

(a) It's clear that out of the Naive Bayes classifiers, the `NB Num` classifier has the best performance -- it has the lowest error of 0.13 and the highest F1 score of 0.87. `NB Num` is based off `x_train_num` which only contains 8 continuous features, of which contains the strongest feature overall (`Weight`). This classifier is only processing continuous data, oppose to `x_train_num_full` and `x_train_cat_full` which contain both continuous and categorical features, whereby these hybrid models reduce efficiency. The test data's continuous distribution also may align with the model very well.

(b) Standardization should improve the performance of the numeric Naive Bayes classifier. Standardization corrects for datasets that have large variance -- large variance is not good, as it means the normal distribution is more spread out decreasing probability density and joint probability.

(c) The drawbacks of fixed-width binning arise if our dataset is sparse, contains several outliers and/or is not distributed evenly. In this event, it's possible that there are heavily populated bins, and potentially bins that aren't populated at all. When applied to a model, this may result in a poor representation of the underlying distribution. A solution is to instead used equal-frequency binning, which ensures all bins are populated relatively evenly, which generally gives better representation.

(d) We want to avoid a false diagnosis of obesity -- this means we want to avoid false positives. In this case we want to focus on getting a solid `Precision` score, which is a metric that measures the prevalence of false positives. We already measured the F1 score, which takes into account `Precision`, however it also takes into account `Recall` which has an inverse relationship with `Precision` (meaning F1 cannot be high unless both `Recall` and `Precision` are high). However, we're only interested in `Precision`, so measuring it directly may be better here.

(e) Yes, the logistic regression has a notably better accuracy and F1 score of around ~10%. Logistic regression is performed on `x_train_norm_full` which has 31 feature columns. 2 of the Naive Bayes classifiers are only being performed on 16 feature columns. Less features means a higher proporition of the columns are 'strong' -- `Height` and `Weight` are present in all the sets, and these are the strongest features. Because all features are weighted evenly, it means there is a drawback in logistic regression here in that these important features have less overall weight in the 31 feature table. Furthermore, too many features may result in overfitting, which will not be good when we want to apply the model to other datasets.

Naive Bayes assumes that the features in the dataset are all conditionally independent, meaning it has a higher bias. However, if the dataset actually follows the bias then Naive Bayes will outperform logistic regression. Naive Bayes also has lower variance. However, the conditionally independent assumption usually isn't ideal. For example, in the obesity dataset Naive Bayes assumes that `Height` and `Weight` are independent -- this is obviously not the case as there is a connection between them. Logistic regression does not suffer from this naive assumption.

(f) Compared to other models, when the dataset is extremely large, gradient descent enables us to train our model much more rapidly. Logistic regression also tends to result in lower error than generative models as the dataset size increases. However, logistic regression does still pertain error properties and so will not always achieve perfect test set performance. The paper "On discriminative vs. generative classifiers: A comparison of logistic regression and naive bayes." by authors Ng, Andrew Y., and Michael I. Jordan explores these errors.